src/mame/drivers/news_r4k.cpp

```diff
@@ -0,0 +1,1309 @@
+	m_hid->irq_out<news_hid_hle_device::MOUSE>().set(FUNC(news_r4k_state::irq_w<KBD>));
+
+	PC8477B(config, m_fdc, 24_MHz_XTAL, pc8477b_device::mode_t::PS2);
+	FLOPPY_CONNECTOR(config, "fdc:0", "35hd", FLOPPY_35_HD, true, floppy_image_device::default_pc_floppy_formats).enable_sound(false);
+
+	// Note - FDC IRQ might go through the FIFO first. Might need to be updated in the future.
+	m_fdc->intrq_wr_callback().set(FUNC(news_r4k_state::irq_w<irq0_number::FDC>));
+	m_fdc->drq_wr_callback().set([this](int status)
+								 { m_fifo0->drq_w<cxd8442q_device::fifo_channel_number::CH2>(status); });
+	m_fifo0->dma_r_cb<cxd8442q_device::fifo_channel_number::CH2>().set([this]()
+																   { return (uint32_t)(m_fdc->dma_r()); });
+
+	DMAC3(config, m_dmac, 0);
+	m_dmac->set_apbus_address_translator(FUNC(news_r4k_state::apbus_virt_to_phys));
+	m_dmac->set_bus(m_cpu, 0);
+	m_dmac->irq_out().set(FUNC(news_r4k_state::irq_w<DMAC>));
+
+	NSCSI_BUS(config, m_scsibus0);
+	NSCSI_BUS(config, m_scsibus1);
+
+	NSCSI_CONNECTOR(config, "scsi0:0", news_scsi_devices, "harddisk");
+	NSCSI_CONNECTOR(config, "scsi0:1", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi0:2", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi0:3", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi0:4", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi0:5", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi0:6", news_scsi_devices, "cdrom");
+
+	NSCSI_CONNECTOR(config, "scsi1:0", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi1:1", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi1:2", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi1:3", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi1:4", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi1:5", news_scsi_devices, nullptr);
+	NSCSI_CONNECTOR(config, "scsi1:6", news_scsi_devices, nullptr);
+
+	// TODO: Actual SPIFI3 clock frequency
+	NSCSI_CONNECTOR(config, "scsi0:7").option_set("spifi3", SPIFI3).clock(16'000'000).machine_config([this](device_t *device)
+																									 {
+																										 spifi3_device &adapter = downcast<spifi3_device &>(*device);
+																										 adapter.irq_handler_cb().set(m_dmac, FUNC(dmac3_device::irq_w<dmac3_device::CTRL0>));
+																										 adapter.drq_handler_cb().set(m_dmac, FUNC(dmac3_device::drq_w<dmac3_device::CTRL0>));
+																									 });
+	NSCSI_CONNECTOR(config, "scsi1:7").option_set("spifi3", SPIFI3).clock(16'000'000).machine_config([this](device_t *device)
+																									 {
+																										 spifi3_device &adapter = downcast<spifi3_device &>(*device);
+																										 adapter.irq_handler_cb().set(m_dmac, FUNC(dmac3_device::irq_w<dmac3_device::CTRL1>));
+																										 adapter.drq_handler_cb().set(m_dmac, FUNC(dmac3_device::drq_w<dmac3_device::CTRL1>));
+																									 });
```

Comment:
There lines are so long as to be bordering on unreadable.  I’d prefer them formatted something like:
```
    // TODO: Actual SPIFI3 clock frequency
    NSCSI_CONNECTOR(config, "scsi0:7").option_set("spifi3", SPIFI3)
            .clock(16'000'000)
            .machine_config(
                [this] (device_t *device)
                {
                    spifi3_device &adapter = downcast<spifi3_device &>(*device);
                    adapter.irq_handler_cb().set(m_dmac, FUNC(dmac3_device::irq_w<dmac3_device::CTRL0>));
                    adapter.drq_handler_cb().set(m_dmac, FUNC(dmac3_device::drq_w<dmac3_device::CTRL0>));
                });
    NSCSI_CONNECTOR(config, "scsi1:7").option_set("spifi3", SPIFI3)
            .clock(16'000'000)
            .machine_config(
                [this] (device_t *device)
                {
                    spifi3_device &adapter = downcast<spifi3_device &>(*device);
                    adapter.irq_handler_cb().set(m_dmac, FUNC(dmac3_device::irq_w<dmac3_device::CTRL1>));
                    adapter.drq_handler_cb().set(m_dmac, FUNC(dmac3_device::drq_w<dmac3_device::CTRL1>));
                });
```

There’s also a trade-off between using `downcast` and `dynamic_cast here – `downcast` will give a more useful error message, but only in `DEBUG=1` builds, so issues may go unnoticed if people only test release builds (which happens far too often); `dynamic_cast` will always blow up, albeit with a less useful message, so it may result in issues being found faster.